### ISSUE TITLE
Errors started coming after updating Linux Mint Cinnamon 19.3 -> 20

### ISSUE THREAD

**Issue by mikkovedru** [autokey/autokey#462]
## Classification:
Bug

## Reproducibility:
Always

## Version
AutoKey version: 0.95.10
Used GUI (Gtk, Qt, or both): both
Installed via: PPA
Linux Distribution: Linux Mint Cinnamon 20
Python (`$ python --version`): Python 3.6.4

## Summary

I have a simple script that opens a directory in the file manager with a hotkey. Example script:
```
# Enter script code
system.exec_command("nemo /media/mikko/Files/Download")
```

In Linux Mint Cinnamon 19.3 (and who knows how many versions before that, because I have been using Mint Cinnamon since 2014 and have been always updating) everything worked without problem. In Mint Cinnamon 20:
1. the script itself works (the directory is opened)
2. but autokey gives an error (both inside the program log as well as trashing my OS notifications list every time I invoke one of my many scripts).

OS Notifications:
![image](https://user-images.githubusercontent.com/2521942/98740227-62e19080-23b3-11eb-8296-361ff3cadafd.png)
![image](https://user-images.githubusercontent.com/2521942/98740342-8c022100-23b3-11eb-857b-3097548ed045.png)

Error Autokey gives:
![image](https://user-images.githubusercontent.com/2521942/98740501-d6839d80-23b3-11eb-97c3-d6c79a4c0bcd.png)

## Actual Results
Here is an attached log file: [2020-11-11_Autokey_error.txt](https://github.com/autokey/autokey/files/5520555/2020-11-11_Autokey_error.txt)

Three things that interesting/worrying:
Errors 1:
```
2020-11-11 00:24:44,862 DEBUG - inotify - Adding watch for /home/mikko/.config/autokey/data/Linux cursors/-button
[2020-11-11 00:24:44,862 pyinotify ERROR] add_watch: cannot watch /home/mikko/.config/autokey/data/Linux cursors/-button WD=-1, Errno=No space left on device (ENOSPC)
2020-11-11 00:24:44,862 ERROR - pyinotify - add_watch: cannot watch /home/mikko/.config/autokey/data/Linux cursors/-button WD=-1, Errno=No space left on device (ENOSPC)
```

Error 2. 
```
(autokey-gtk:305378): Gdk-CRITICAL **: 00:24:45.478: gdk_window_thaw_toplevel_updates: assertion 'window->update_and_descendants_freeze_count > 0' failed
2020-11-11 00:24:45,986 ERROR - interface - Got BadWindow error while requesting window information.
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/autokey/interface.py", line 1074, in get_window_info
    return self._get_window_info(window, traverse)
  File "/usr/lib/python3/dist-packages/autokey/interface.py", line 1080, in _get_window_info
    new_wm_title = self._try_get_window_title(window)
  File "/usr/lib/python3/dist-packages/autokey/interface.py", line 1128, in _try_get_window_title
    atom = self._try_read_property(window, self.__VisibleNameAtom)
  File "/usr/lib/python3/dist-packages/autokey/interface.py", line 1145, in _try_read_property
    return window.get_property(property_name, 0, 0, 255)
  File "/usr/lib/python3/dist-packages/Xlib/xobject/drawable.py", line 455, in get_property
    r = request.GetProperty(display = self.display,
  File "/usr/lib/python3/dist-packages/Xlib/protocol/rq.py", line 1369, in __init__
    self.reply()
  File "/usr/lib/python3/dist-packages/Xlib/protocol/rq.py", line 1389, in reply
    raise self._error
Xlib.error.BadWindow: <class 'Xlib.error.BadWindow'>: code = 3, resource_id = <class 'Xlib.xobject.resource.Resource'>(0x0301eeb5), sequence_number = 3342, major_opcode = 20, minor_opcode = 0
2020-11-11 00:24:46,038 DEBUG - interface - Grabbing hotkey: ['<super>'] 't'
2020-11-11 00:24:46,038 DEBUG - interface - Grabbing hotkey: ['<super>'] 'r'
2020-11-11 00:24:46,038 DEBUG - interface - Grabbing hotkey: ['<super>'] 't'
2020-11-11 00:24:46,039 DEBUG - interface - Grabbing hotkey: ['<super>'] 'r'
2020-11-11 00:24:46,039 DEBUG - interface - Grabbing hotkey: ['<super>'] 't'
2020-11-11 00:24:46,039 DEBUG - interface - Grabbing hotkey: ['<super>'] 'r'
2020-11-11 00:24:46,039 DEBUG - interface - Grabbing hotkey: ['<super>'] 't'
2020-11-11 00:24:46,039 DEBUG - interface - Grabbing hotkey: ['<super>'] 'r'
X protocol error:
<class 'Xlib.error.BadWindow'>: code = 3, resource_id = <class 'Xlib.xobject.resource.Resource'>(0x0301eeb5), sequence_number = 3354, major_opcode = 33, minor_opcode = 0
X protocol error:
<class 'Xlib.error.BadWindow'>: code = 3, resource_id = <class 'Xlib.xobject.resource.Resource'>(0x0301eeb5), sequence_number = 3355, major_opcode = 33, minor_opcode = 0
X protocol error:
<class 'Xlib.error.BadWindow'>: code = 3, resource_id = <class 'Xlib.xobject.resource.Resource'>(0x0301eeb5), sequence_number = 3356, major_opcode = 33, minor_opcode = 0
X protocol error:
<class 'Xlib.error.BadWindow'>: code = 3, resource_id = <class 'Xlib.xobject.resource.Resource'>(0x0301eeb5), sequence_number = 3357, major_opcode = 33, minor_opcode = 0
```

Error 3. The actual error caused by me pressing the hotkey.
```
2020-11-11 00:24:52,462 DEBUG - iomediator - Key.CONTROL pressed
2020-11-11 00:24:52,477 DEBUG - iomediator - Key.ALT pressed
2020-11-11 00:24:52,703 DEBUG - service - Raw key: 'd', modifiers: [<Key.ALT: '<alt>'>, <Key.CONTROL: '<ctrl>'>], Key: d
2020-11-11 00:24:52,703 DEBUG - service - Window visible title: 'autokey-gtk --verbose', Window class: 'gnome-terminal-server.Gnome-terminal'
2020-11-11 00:24:52,704 INFO - service - Matched Script "Open Download directory" with hotkey and prompt=False
2020-11-11 00:24:52,704 DEBUG - service - Script runner executing: Script('Open Download directory')
2020-11-11 00:24:52,705 ERROR - service - Ignored locking error in handle_keypress
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/autokey/service.py", line 207, in __tryReleaseLock
    self.configManager.lock.release()
RuntimeError: release unlocked lock
2020-11-11 00:24:52,766 ERROR - service - Script error
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/autokey/service.py", line 485, in execute
    exec(script.code, scope)
  File "<string>", line 2, in <module>
  File "/usr/lib/python3/dist-packages/autokey/scripting.py", line 497, in exec_command
    if output[-1] == "\n":
IndexError: string index out of range
2020-11-11 00:24:52,830 DEBUG - iomediator - Key.ALT released
2020-11-11 00:24:52,837 DEBUG - iomediator - Key.CONTROL released
```

**Comment by josephj11** [comment]
We've seen this error message a number of times before. Not sure, but I think the last time we saw this, it had something to do with an update to X11 from 19.x to 20.x.

**Comment by mikkovedru** [comment]
According to [wikipedia](https://en.wikipedia.org/wiki/Linux_Mint_version_history#Version_history_of_common_programs), the X.Org Server's versions are:
- 1.19.6 for Mint 19.1 (and I assume for 19.3 as well)
- 1.20.8 for Mint 20

**Comment by johncarterofmars** [comment]
I'm using Mint 20 and have no errors. I have been using AutoKey since Mint 19.1. So if there is something I can do to help test, I am happy to do so. I do have some minor issues but not with the overall use of the app.

**Comment by josephj11** [comment]
While this is a valid issue, we might do better to discuss this on [Gitter](https://gitter.im/autokey/autokey) where more eyes will see it.

**Comment by alanesq** [comment]
Hi,
Did you find a fix for this as I am experiencing it after upgrading my Linux Mint version

**Comment by josephj11** [comment]
A few thoughts.

Error 1:
No idea, but the easiest way to break shell scripts is to use directory names with white space or special characters in their names. Try renaming `Linux cursors` to `Linux_cursors` and see if that helps. Do that from inside AutoKey rather than at the command line, just to be safe.

Error 2:
Probably the xlib version update - no idea how to deal with that.

Error 3:
The AutoKey error comes from looking at output[-1]. The only way that can be out of range is if output is an empty string. I think I have seen a similar bug discussed and fixed by the devs in our develop branch.

It looks like it's the result of a recent change we made to trim a trailing newline from the output of a command. But if the command has no output, it causes this error. Your particular command does not produce any output on stdout so it hits this error.

Since we don't currently have a path to a new release, there are a couple of ways to proceed.

1) Add something to your command that produces output.
    You could replace the direct call to Nemo with a call to a bash script which starts with `echo ""` before you invoke Nemo. You would have to do this for any similar command that produces no output.


2) Patch line 497 of your copy of scripting.py, changing it from

`if output.endswith("\n"):`

to something like (untested)

`if len(output) > 0 and output.endswith("\n"):`

Assuming my code is correct (I'm a Python beginner), this would fix the problem itself, so you wouldn't have to worry about it in the future.

**Comment by alanesq** [comment]
> `if len(output) > 0 and output.endswith("\n"):`
> Assuming my code is correct (I'm a Python beginner), this would fix the problem itself, so you wouldn't have to worry about it in the future.

Thanks very much for your reply, I have tried modifying the code as you suggested and this seems to have resolved the problem :-)

**Comment by BlueDrink9** [comment]
This ought to be closed as a duplicate of #483 @josephj11
